prow/plugins/trigger/ic.go

```diff
@@ -159,7 +159,19 @@ func handleIC(c client, trustedOrg string, ic github.IssueCommentEvent) error {
 
 	var errors []error
 	for _, job := range requestedJobs {
-		if !job.RunsAgainstBranch(pr.Base.Ref) || (!job.AlwaysRun && job.RunIfChanged != "" && !job.RunsAgainstChanges(changedFiles)) {
+		build := true
+		if !job.RunsAgainstBranch(pr.Base.Ref) {
+			build = false
+		} else if !job.AlwaysRun && job.RunIfChanged != "" {
+			changes, err := files()
+			if err != nil {
+				return err
+			}
+			if !job.RunsAgainstChanges(changes) {
+				build = false
```

Comment:
It will only be silent if SkipReport is true, otherwise it marks the context as "skipped".  If a job doesn't match the regex it should be marked skipped and should not actually run, marking the job "skipped" **is** honoring the user's request.

If we got rid of this conditional and always ran requested RunIfChanged job without checking that the changes match the regex a `/test all` would cause all RunIfChanged jobs to be triggered even if no changed files match the regexes.

Reply:
I'm not opposed to changing the behavior of these jobs to allow them to be explicitly triggered on PRs that they shouldn't normally be run on, but we would need to change how `/test all` works because currently `/test all` and `/test job` are both considered explicit triggers of job `job`.

We could modify all the rerun command regexes in `config.yaml` to not match `/test all` and instead treat a `/test all` as a special case more like an `/ok-to-test`.

Reply:
Hmm, good points. In some sense it seems simpler and ok if `/test all` triggers tests that allow for `all` even if they're not running on a PR changing specific files. I guess the way I thought of it was - `/test` is an explicit user action and we should honor that. The automatic "run if changed" behavior is nice to catch PRs touching specific files without needing human intervention but isn't saying "don't run if not changed." As a human if I've determined that I want to run a test and tell the bot to do it, it should listen. If we want a `/test-i-really-mean-it` for that case, maybe that would be better. If changing this behavior would be a large regression in the current behavior, I agree that we shouldn't change it in this PR.

Reply:
Changing the behavior as you suggest would be nice, but I think there are some complications that we need to consider first. 
Specifically, if a RunIfChanged job is manually run on a PR that doesn't require the job to be run (via explicit `/test job` or `/test all`) and the job fails, how could the PR author mark the job as skipped in order to ignore the results of the job and merge the PR?
Having a `/test-i-really-mean-it` or `/test-no-skip` command would allow us to differentiate between requests to mark a job as skipped and requests to actually run the job, but I think we should make that a separate PR if we want to do that.

Reply:
Hmm, and I think your comment is also getting at a deeper question that we've skirted around a little bit in other issues -- if someone manually triggers a job for a PR, should it become sticky and be required for that PR to merge in the future? In any case, I agree that these changes should land later.

Reply:
Trying to catch up here, so if I have a job that is SkipReport false, run_if_changed and I do `/test job` will it run and post results? 
Also `/test all` is only enforced via a presubmit requiring this to be in the regex for jobs with alwaysrun: true, so I'm not sure why this is brought up here?

Reply:
> Specifically, if a RunIfChanged job is manually run on a PR that doesn't require the job to be run (via explicit /test job or /test all) and the job fails, how could the PR author mark the job as skipped in order to ignore the results of the job and merge the PR?

By pushing a new commit? Or perhaps we could have an admin/owner command to clear a job as skipped?
If I do `/test pull-kubernetes-cross` on someone's PR and it fails, I do not want this PR to merge that commit. Handling what happens when they push a new commit seems a bit debateable though.

Reply:
A skip command that clears the context sounds useful, especially if tide continues to look at the PR status vs individual contexts.

Reply:
>Also /test all is only enforced via a test-infra presubmit requiring this to be in the regex for jobs with AlwaysRun: true, so I'm not sure why this is brought up here?

A job being triggered by `/test all` and `/test job` are currently indistinguishable if the job's regex matches `/test all`. So if we considered `/test job` to mean "force run the job" `/test all` would force run all jobs including all RunIfChanged jobs.
Are you suggesting that we change the regexes for RunIfChangedJobs to never match `/test all` and enforce that with a presubmit? That could work in combination with a `/skip job` command that marks a job's github status context as 'skipped' if the job does not need to be run on the PR .

Reply:
> Are you suggesting that we change the regexes for RunIfChangedJobs to never match /test all and enforce that with a presubmit? That could work in combination with a /skip job command that marks a job's github status context as 'skipped' if the job does not need to be run on the PR .

I do like the idea of having `/test all` just be a prow method for "test all jobs that should run against this PR", but what I meant is precisely that /test all and /test job are the same, but what I meant is that right now we have a presubmit that only allows `/test all` to match alwaysRun: true jobs (and it requires that for them).

But regardless if I manually request a test on a PR it should run regardless of run_if_changed (the current behavior) and post the status but only once per request comment, otherwise we have a behavior change that needs further discussion..

Reply:
>what I meant is that right now we have a presubmit that only allows /test all to match alwaysRun: true jobs

Got it.
> if I manually request a test on a PR it should run regardless of run_if_changed (the current behavior)

This isn't the behavior anymore. That was previously the behavior because the issue comment part of trigger didn't know about RunIfChanged at all. Now the job would just be marked as skipped if the PR doesn't match the RunIfChanged regex.
I agree that we should change the behavior so that the job is run if it is explicitly triggered, but should we make sure that we have something like `/skip job` before we do that? The temporary alternative would be to handle the case where we would want to clear the status of the manually run job by having the PR author push a new commit.

Reply:
> The temporary alternative would be to handle the case where we would want to clear the status of the manually run job by having the PR author push a new commit.

I think this is fine, this puts us back to where things where before. Hopefully the new commit actually fixes the job. I'd like to at least get to this point so we can improve the health of the cross build with the release coming around, but I also think we should hash out plans for something like  a skip command next. 
I think we can afford to go ahead and get "just push a new commit" working now and make the skip solution a follow-up though.